File: codes/exp3/semipar_SNTPS_tmbstan.R
rm(list = ls())
setwd("C:/Users/joaquin/Desktop/new_models")
library(TMB)
library(mgcv)
library(sn)
library(MASS)
library(tidyverse)
library(gridExtra)
library(grid)
library(lattice)
library(tmbstan)
library(rstan)
library(tmbstan)
library(rstan)
library(bayesplot)
library(geoR)
library(raster)
library(Matrix)


options(scipen=999)

#=================================================
# Compilamos el modelo y lo cargamos en R
TMB::compile('semipar_SNTPS_tmbstan.cpp')
dyn.load(dynlib("semipar_SNTPS_tmbstan"))
# #=================================================
# 


#=============================================================================
#                              Applied analysis
#=============================================================================
data = read.csv("north2.csv", header = T)
head(data, 3)

# data <- data[sample(nrow(data), 500), ]
# dim(data)


#====================================================
#      Obtenemos matriz penalizada desde mgcv
#====================================================
# Creamos una variable artificial
z <- rep(0, length(data$cpue))
data$z <- z
head(data)


# data <- sample_n(data, 2500)
# dim(data)

#===================
#     mgcv setup
#===================
m0 <- gam(z ~ year + trim + destine + depth  + s(latitude, longitude, bs = "tp", k = 13), data = data, fit = FALSE)
#m0 <- gam(z ~ year + depth  + s(latitude, longitude, bs = "tp", k = 13), data = data, fit = FALSE) 


Xtps <- m0$X[, c(-1, -2, -3, -4, -5)]                    # Matricial form without intercept and the parameters asociated with the covariates
#Xtps <- m0$X[, c(-1, -2. -3)]                    # Matricial form without intercept and the parameters asociated with the covariates
Stps <- m0$smooth[[1]]$S[[1]]                # Extrtact penelization matrices
dim(Stps)

S_list = list(Stps)
S_combined = .bdiag(S_list)                               # join S's in sparse matrix
Sdims = unlist(lapply(S_list,nrow))                       # Find dimension of each S (in this case only one)
#----------------------------------------

#===========================================================================
# For report, used for constructing plots
latitude = seq(min(data$latitude), max(data$latitude), by = 0.1)
longitude = seq(min(data$longitude), max(data$longitude), by = 0.1)

# by cambia para hacer coincidir los largos de los vectores
tpsReport = PredictMat(m0$smooth[[1]], data = data.frame(latitude, longitude))
tpsReport = list(tpsReport)



#-------------------------------------------



# # Plot
# to_plot = TRUE   # generate plots
# 
# class(data)
# 
# if (to_plot){
#   ggplot(data) +
#     geom_point(aes(x = longitude, y = latitude, fill = cpue)) +
#     coord_equal() +
#     scale_fill_viridis_c()
# }





#=================================================================
#                           TMB modelling
#=================================================================
data$year <- as.factor(data$year)
data$trim <- as.factor(data$trim)
data$destine <- as.factor(data$destine)


#======================================
#                TMB data
#======================================
tmb_data = list(likelihood = 3,
                #cpue     = as.vector(data$cpue),          # Response
                cpue     = sqrt(as.vector(data$cpue)),          # Response
                year = as.numeric(data$year) - 1,
                trim = as.numeric(data$trim) - 1,
                destine = as.numeric(data$destine) - 1,
                depth = as.numeric(data$depth),
                #X    = model.matrix(~1 + data$depth + as.numeric(data$year) - 1),
                TPS     = Xtps,                                   # Design matrix, without intercept and betas
                S     = as(S_combined, "dgTMatrix"),                          # Combined penalty matrix
                Sdims = Sdims,
                tpsReport = .bdiag(tpsReport))



#=====================================
#            TMB parameters
#=====================================
tmb_par = list(beta0 = 0.1,
               beta_year = rep(0.1, length(levels(data$year))),
               beta_trim = rep(0.1, length(levels(data$trim))),
               beta_destine = rep(0.1, length(levels(data$destine))),
               beta_depth = 0.1,
               smoothCoefs = rep(rnorm(ncol(Xtps), 0, 1)),              # Spline coefficients
               logalpha = exp(0.1),
               #logalpha = rep(rep(0.1,length(Sdims))),       # Log spline penalization coefficients
               logsigma = exp(0.1),
               loglambda = exp(0.1))
               




#=====================================
#             SKEW NORMAL MODEL
#=====================================
obj <- MakeADFun(tmb_data, random = c("smoothCoefs"), tmb_par, DLL="semipar_SNTPS_tmbstan")


# Lognormal
startTime <- Sys.time()
opt <- nlminb(obj$par,obj$fn,obj$gr)
rep <- sdreport(obj)
endTime <- Sys.time()
timeUsed = endTime - startTime
print(timeUsed)

# Gamma
startTime <- Sys.time()
opt2 <- nlminb(obj$par,obj$fn,obj$gr)
rep2 <- sdreport(obj)
endTime <- Sys.time()
timeUsed = endTime - startTime
print(timeUsed)


# Skew-normal
startTime <- Sys.time()
opt3 <- nlminb(obj$par,obj$fn,obj$gr)
rep3 <- sdreport(obj)
endTime <- Sys.time()
timeUsed = endTime - startTime
print(timeUsed)

obj$report()$log_lik
#-------------------------------------------

library(TMBhelper)
TMBAIC(opt)
TMBAIC(opt2)
TMBAIC(opt3)


#===========================================================================================
#                                        Fit with tmbstan
#===========================================================================================
# Calculate the number of cores
no_cores <- parallelly::availableCores()  - 1

init.fn <- function()
  split(unname(obj$env$last.par.best),names(obj$env$last.par.best))


startTime <- Sys.time()
fit = tmbstan(obj, chains = 3, open_progress = FALSE, 
              init = init.fn, control = list(max_treedepth = 13, adapt_delta = 0.9), 
              iter = 3000, warmup=1000, cores = no_cores, seed=483892929)
endTime <- Sys.time()
timeUsed = endTime - startTime
print(timeUsed)


## 1.9 hours minutes!!!


s <- summary(fit, probs = c(0.25, 0.75))
s$summary  # all chaines merged



setwd("C:/Users/joaquin/Desktop/new_models/output_real_application")

posterior_tps_tmbstan <- as.matrix(fit)

saveRDS(posterior_tps_tmbstan, file='posterior_tps_tmbstan.RDS')
saveRDS(fit, file='fit_tps_tmbstan.RDS')

c_light <- c("#DCBCBC")
c_light_highlight <- c("#C79999")
c_mid <- c("#B97C7C")
c_mid_highlight <- c("#A25050")
c_dark <- c("#8F2727")
c_dark_highlight <- c("#7C0000")


traceplot(fit, pars=names(obj$par), inc_warmup=TRUE)

#load package
require(MCMCvis)
MCMCsummary(fit, round = 2)

#pairs(fit, pars=names(fit)[-grep('beta0',names(fit))][1:10])
pairs(fit, pars = c("logsigma", "logalpha", "lp__"), las = 1) # below the diagonal
pairs(fit, pars = c("logsigma", "logalpha", "loglambda")) # below the diagonal


pairs(fit, pars = c("logsigma", "logalpha", "lp__"), las = 1) # below the diagonal


params_cp <- as.data.frame(fit)
names(params_cp) <- gsub("chain:1.", "", names(params_cp), fixed = TRUE)
names(params_cp) <- gsub("[", ".", names(params_cp), fixed = TRUE)
names(params_cp) <- gsub("]", "", names(params_cp), fixed = TRUE)
params_cp$iter <- 1:6000


# logalpha
par(mfrow=c(3,2),mar=c(4,4,0.5,0.5), oma=c(2,3,1,1))
plot(params_cp$iter, params_cp$logalpha, col=c_dark, pch=16, cex=0.8, type = "l",
     xlab="Iteration", ylab="logalpha", cex.lab=1.3, cex.axis=1.3)

running_means_alpha = sapply(params_cp$iter, function(n) mean(params_cp$logalpha[1:n]))
plot(params_cp$iter, running_means_alpha, col=c_dark, pch=16, cex=0.8,  cex.lab=1.3, cex.axis=1.3,
     xlab="Iteration", ylab="MCMC mean of logalpha")
abline(h=mean(running_means_alpha), col="grey", lty="dashed", lwd=3)



# logsigma
plot(params_cp$iter, params_cp$logsigma, col=c_dark, pch=16, cex=0.8, type = "l",
     xlab="Iteration", ylab="logsigma",  cex.lab=1.3, cex.axis=1.3)

running_means_sigma = sapply(params_cp$iter, function(n) mean(params_cp$logsigma[1:n]))
plot(params_cp$iter, running_means_sigma, col=c_dark, pch=16, cex=0.8,  cex.lab=1.3, cex.axis=1.3,
     xlab="Iteration", ylab="MCMC mean of logsigma")
abline(h=mean(running_means_sigma), col="grey", lty="dashed", lwd=3)


# loglambda
plot(params_cp$iter, params_cp$loglambda, col=c_dark, pch=16, cex=0.8, type = "l",
     xlab="Iteration", ylab="loglambda",  cex.lab=1.3, cex.axis=1.3)

running_means_lambda = sapply(params_cp$iter, function(n) mean(params_cp$loglambda[1:n]))
plot(params_cp$iter, running_means_lambda, col=c_dark, pch=16, cex=0.8,  cex.lab=1.3, cex.axis=1.3,
     xlab="Iteration", ylab="MCMC mean of loglambda")
abline(h=mean(running_means_lambda), col="grey", lty="dashed", lwd=3)
mtext("Convergence of the parameters alpha, sigma and lambda", outer=TRUE,  cex=1, line=-0.5)




# Divergences
divergent = get_sampler_params(fit, inc_warmup=FALSE)[[1]][,'divergent__']
sum(divergent)

## Methods provided by 'rstan'
class(fit)
methods(class ="stanfit")

get_cppo_mode(fit)
get_stancode(fit)
get_stanmodel(fit)
log_prob(fit)
loo(fit)
traceplot(fit)

#launch_shinystan(fit)

## ESS and Rhat from rstan::monitor
mon = monitor(fit)
max(mon$Rhat)
min(mon$Tail_ESS)

# evalaute problem of convergence
sum(mon$Rhat > 1.01)
sum(mon$Tail_ESS < 400)

source('monitornew.R')
source('monitorplot.R')
source('stan_utility.R')

which_min_ess = which.min(mon[1:200, 'Tail_ESS'])
plot_local_ess(fit = fit, par = which_min_ess, nalpha = 10)

plot_quantile_ess(fit = fit, par = which_min_ess, nalpha = 50)

plot_change_ess(fit = fit, par = which_min_ess)

check_rhat(fit)
check_treedepth(fit, 10)
check_energy(fit)  #
check_div(fit)


# Variance
library(MCMCvis)
color_scheme_set("viridisE")
#mcmc_rank_hist(fit, pars = c("sigma_beta_year", "sigma_beta_depth", "sigma_beta_trim", "sigma_beta_destine"), ref_line = TRUE)
mcmc_rank_hist(fit, pars = c("logalpha", "logsigma", "loglambda"), ref_line = TRUE)   # spatial random field


## Extract marginal posteriors
posterior <- as.matrix(fit)

exp(mean(posterior[, "logalpha"]))
exp(mean(posterior[, "logsigma"]))
exp(mean(posterior[, "loglambda"]))

exp(opt$par[3:5])





# Simulating from the SIMULATE function
mat_sim = matrix(data=NA, nrow=length(obj$simulate()$cpue_sim), ncol=100)
mat_sim


for(j in 1:ncol(mat_sim)){
  for(i in 1:nrow(mat_sim)){
    mat_sim[, j] = obj$simulate()$cpue_sim
  }
}
mat_sim

# Plot TPS model using Gamma likelihood
hist(data$cpue, col = "gray90", prob = TRUE, main = "100 simulated samples", cex.main = 2, xlab = "", col.lab = 'blue', col.main="blue", cex.lab = 1.4, cex.axis = 1.2, ylim = c(0, 0.008))
for (j in 1: ncol(mat_sim)){
  lines(x = density(x = mat_sim[, j]),  lty="dotted", col="azure4", lwd=1)
}
#legend("topright", "A", bty = "n", text.col="blue", cex = 1.4)
legend(6, 0.3, c("response", "simulations"), lwd=4, col=c("gray90", "azure4"), cex = 1.5)

File: codes/exp3/semipar_SNTPS_tmbstan.cpp
// include libraries
#include <TMB.hpp>
#include <Eigen/Sparse>
#include <vector>
#include <string>
using namespace density;
using Eigen::SparseMatrix;
// Can also choose which likelihood to use.
// Lognormal density
template<class Type>
Type dlognorm(Type x, Type meanlog, Type sdlog, int give_log=0){
  //return 1/(sqrt(2*M_PI)*sd)*exp(-.5*pow((x-mean)/sd,2));
  Type logres = dnorm( log(x), meanlog, sdlog, true) - log(x);
  if(give_log) return logres; else return exp(logres);
}
// Inverse gamma
template<class Type>
Type dinvgauss(Type x, Type mean, Type shape, int give_log=0){
  Type logres = 0.5*log(shape) - 0.5*log(2*M_PI*pow(x,3)) - (shape * pow(x-mean,2) / (2*pow(mean,2)*x));
  if(give_log) return logres; else return exp(logres);
}
// dcauchy for hyperparameters
template<class Type>
Type dcauchy(Type x, Type mean, Type shape, int give_log=0){
  Type logres = 0.0;
  logres-= log(M_PI);
  logres-= log(shape);
  // Note, this is unstable and should switch to log1p formulation
  logres-= log(1 + pow( (x-mean)/shape ,2));
  if(give_log) return logres; else return exp(logres);
}


template<class Type>
Type objective_function<Type>::operator() ()
{
  
  
  //============================================
  //                 DATA
  //============================================
  DATA_INTEGER(likelihood);                    // the likelihood function to use
  //DATA_INTEGER(model);                    // the likelihood function to use
  DATA_VECTOR(cpue);                              // The response
  
  DATA_FACTOR(year);             // year as a factor
  DATA_FACTOR(trim);             // trimester
  DATA_FACTOR(destine);          // destine
  
  DATA_VECTOR(depth);
  
  
   
  DATA_MATRIX(TPS);                              // Design matrix for splines
  
  DATA_SPARSE_MATRIX(S);                       // Penalization matrix diag(S1,S2,S3,S4,S5) without storing off-diagonal zeros.
  DATA_IVECTOR(Sdims);                         // Dimensions of S1,S2,S3,S4 and S5
  DATA_SPARSE_MATRIX(tpsReport);               // Design matrix for report of splines
  
  
  //----------------------------------
  
  //============================================
  //                 PARAMETERS
  //============================================
  
  // Parameters for factors  
  //PARAMETER_VECTOR(beta);	        // beta for trimester  (factor)
  PARAMETER(beta0);
  PARAMETER_VECTOR(beta_year);
  PARAMETER_VECTOR(beta_trim);
  PARAMETER_VECTOR(beta_destine);
  
  PARAMETER(beta_depth);
  
  PARAMETER_VECTOR(smoothCoefs);               // Spline regression parameters
  //PARAMETER_VECTOR(logalpha);                  // Penalization parameters
  PARAMETER(logalpha);                  // Penalization parameters
  PARAMETER(logsigma);                        // Penalizatiion parameter
  
  PARAMETER(loglambda);                        // skew normal parameter (slant)
  
  

  //==========================================
  // Transformed parameters
  Type sigma  = exp(logsigma);
  //Type lambda = exp(loglambda);
  //vector<Type> alpha = exp(logalpha);
  Type alpha = exp(logalpha);
  Type lambda = exp(loglambda);
  
  //==========================================  
  // Priors
  Type nlp = Type(0.0);                                 // negative log prior  (priors)
  
  //nlp-= dnorm(beta,    Type(0.0), Type(10.0), true).sum();
  nlp-= dnorm(beta0,        Type(0.0), Type(5.0), true);
  nlp-= dnorm(beta_year,    Type(0.0), Type(5.0), true).sum();
  nlp-= dnorm(beta_trim,    Type(0.0), Type(5.0), true).sum();
  nlp-= dnorm(beta_destine, Type(0.0), Type(5.0), true).sum();
  
  nlp-= dnorm(beta_depth,   Type(0.0), Type(5.0), true);
  
  // Variance component
  nlp-= dcauchy(sigma,   Type(0.0),   Type(2.0));
  // 
  // Penalty parameter
  //nlp-= dgamma(alpha,   Type(1.0),   Type(2.0), true).sum(); ---> WORKS!!
  
  //nlp-= dcauchy(alpha,   Type(1.0),   Type(2.0)).sum();
  
  nlp-= dnorm(lambda, Type(0.0),   Type(1.0), true);
  
  nlp-= dnorm(alpha, Type(0.0),   Type(1.0), true);
  
  // 
  // Slant parameter of Skew Normal
  //nlp-= dnorm(lambda,   Type(0.0),  Type(10.0));
  
  
  // Prior for smoothCoefs
  //Type penalty = sigma/alpha;
  
  nlp-= dnorm(smoothCoefs, Type(0.0), Type(1.0), true).sum(); 
  
  //==========================================
  // OBJETIVE FUNCTION
  Type nll=0;
  
  int k=0;  // Counter
  for(int i=0;i<Sdims.size();i++){
    int m_i = Sdims(i);
    vector<Type> smoothCoefs_i = smoothCoefs.segment(k,m_i);       // Recover betai
    SparseMatrix<Type> S_i = S.block(k,k,m_i,m_i);  // Recover Si
    //nll -= Type(0.5)*m_i*logalpha(i) - Type(0.5)*alpha(i)*GMRF(S_i).Quadform(smoothCoefs_i);
    //nll -= Type(0.5)*m_i*logalpha - Type(0.5)*alpha*GMRF(S_i).Quadform(smoothCoefs_i);
    nll -= Type(0.5)*m_i*logalpha - Type(0.5)*alpha*GMRF(S_i).Quadform(smoothCoefs_i);
    k += m_i;
  }
  
  
  
  
  
  // We create a vector of means
  //mu = beta0 + beta_year(year) + beta_depth*depth + X*smoothCoefs;
  
  vector<Type> mu(cpue.size());
  //mu = X*beta + TPS*smoothCoefs;
  //mu = beta0 + beta_year(year) + beta_trim(trim) + beta_destine(destine) + beta_depth*depth + TPS*smoothCoefs;

  mu = beta0 + beta_year(year) + beta_trim(trim) + beta_destine(destine) + beta_depth*depth + TPS*smoothCoefs;
  //mu = exp(beta0 + beta_year(year) + beta_depth*depth + TPS*smoothCoefs);
  
  //===============================================================
  // log-likelihood for the response
  
  // int n = cpue.size();
  // vector<Type> z = cpue - mu;  
  // for( int i = 0; i<n ; i++){
  //   // Skew-Normal case
  //   if(likelihood==1)
  //   nll -= dsn(z(i)/sigma, lambda, true) - log(sigma);
  //   // Normal case
  //   if(likelihood==2)
  //   nll -= dlognorm(cpue(i), mu(i), sigma, true);
  //   if(likelihood==3)
  //   nll -= dgamma(cpue(i), 1/pow(sigma,2), exp(mu(i))*pow(sigma,2), true);
  // }
  
// Probability of the data, given random effects (likelihood)
  int n = cpue.size();
  vector<Type> log_lik(n);
  vector<Type> z = cpue - mu;  
  
  for( int i = 0; i< n; i++){
    if(likelihood==1) // lognormal case
      log_lik(i) = dlognorm(cpue(i), mu(i), sigma, true);
    if(likelihood==2) // gamma case
      log_lik(i) = dgamma(cpue(i), 1/pow(sigma,2), exp(mu(i))*pow(sigma,2), true);
    if(likelihood==3) // gamma case
      log_lik(i) = dsn(z(i)/sigma, lambda, true) - log(sigma);
  }
  nll = -log_lik.sum(); // total NLL  
  
  //---------------------------------------------------------------
  
  
  // Simule data from the mu (only an experimental analysis because TMB not support rdsn)
  // Simule data from the mu
  vector<Type> cpue_sim(n);
  for( int i=0; i<n; i++){
    //if(likelihood==2)
    //   SIMULATE {
    //     cpue_sim(i) = rnorm(log(mu(i)), sigma);
    //     REPORT(cpue_sim)};
    // if(likelihood==3)
      SIMULATE {
        cpue_sim(i) = rgamma(1/pow(sigma,2), mu(i)*pow(sigma,2));
        REPORT(cpue_sim);
      }
  }
  //----------------------------------
  
  
  
  //==========================================
  // Derived quantities
  vector<Type> preds = mu;
  vector<Type> splines2D = tpsReport*smoothCoefs;
  
  // Jacobian adjustment for transformed parameters
  nll -= logsigma + logalpha + loglambda;   // add logalpha? how?
  
  // Calculate joint negative log likelihood
  Type jnll = nll + nlp;
  
  
  //=====================================================================================================
  // REPORT
  REPORT(nll);
  
  //REPORT(beta);
  REPORT(beta0);
  REPORT(beta_year);
  REPORT(beta_depth);
  //REPORT(logsigma);
  //REPORT(loglambda);
  REPORT(logalpha);
  REPORT(smoothCoefs);
  REPORT(mu);
  REPORT(log_lik);
  REPORT(splines2D);
  
  //=====================================================================================================
  // AD REPORT
  ADREPORT(beta0);		            
  ADREPORT(logsigma);
  //ADREPORT(loglambda);	
  //ADREPORT(logalpha); 
  ADREPORT(smoothCoefs);
  ADREPORT(splines2D);
  return jnll;
}